banzai_nres/rv.py

```diff
@@ -0,0 +1,133 @@
+    # The time given in the NRES header is the exposure start time; correct this to the midpoint
+    # Eventually, we will want to use the flux-weighted mid-point calculated from emeter data
+    obs_time = Time(time, location=site_location) + exptime / 2.0 * units.s
+    barycorr_rv = sky_coordinates.radial_velocity_correction(obstime=obs_time)
+    bjd_tdb = obs_time.tdb + obs_time.light_travel_time(sky_coordinates)
+    return barycorr_rv.to('m/s').value, bjd_tdb.to_value('jd')
+
+
+def cross_correlate_over_traces(image, orders_to_use, velocities, template):
+    ccfs = []
+
+    for i in orders_to_use:
+        logger.info(f'Cross correlating for order: {i}', image=image)
+        order = image.spectrum[np.logical_and(image.spectrum['fiber'] == image.science_fiber,
+                                              image.spectrum['order'] == i)][0]
+
+        # Only pass in the given wavelength range +- 1 Angstrom to boost performance
+        order_indices = np.logical_and(template['wavelength'] >= np.min(order['wavelength'] - 1.0),
+                                       template['wavelength'] <= np.max(order['wavelength'] + 1.0))
+        x_cor = cross_correlate(velocities, order['wavelength'], order['flux'], order['uncertainty'],
+                                template['wavelength'][order_indices], template['flux'][order_indices])
+        ccfs.append({'order': i, 'v': velocities, 'xcor': x_cor})
+    return Table(ccfs)
+
+
+class RVCalculator(Stage):
+    TEMPLATE_FILENAME = pkg_resources.resource_filename('banzai_nres', 'data/g2v_template.fits')
+
+    def do_stage(self, image) -> ObservationFrame:
+        # Load in the template
+        template_hdu = fits.open(self.TEMPLATE_FILENAME)
+
+        template = {'wavelength': template_hdu[1].data['wavelength'], 'flux': template_hdu[1].data['flux']}
+        # This is mostly arbitrary. Just pick orders near the center of the detector
+        orders_to_use = np.arange(75, 101, 1)
+
+        # for steps in 1 km/s from -2000 to +2000 km/s
+        velocities = np.arange(-2000, 2001, 1)
+
+        coarse_ccfs = cross_correlate_over_traces(image, orders_to_use, velocities, template)
+
+        # take the peak
+        velocity_peaks = np.array([velocities[np.argmax(ccf['xcor'])] for ccf in coarse_ccfs])
+        best_v = stats.sigma_clipped_mean(velocity_peaks, 3.0)
+        velocities = np.arange(best_v - 2, best_v + 2 + 1e-4, 1e-3)
+
+        final_ccfs = cross_correlate_over_traces(image, orders_to_use, velocities, template)
+
+        rvs_per_order = np.array([ccf['v'][np.argmax(ccf['xcor'])] for ccf in final_ccfs])
+        # Calculate the peak v (converting to m/s) in the spectrograph frame
+        rv_measured = stats.sigma_clipped_mean(rvs_per_order, 3.0) * 1000
+
+        # TODO: Add ra and dec to observation frame object
+        # Compute the barycentric RV and observing time corrections
+        rv_correction, bjd_tdb = barycentric_correction(image.dateobs, image.exptime,
```

Comment:
barycentric_correction returns a jd I think, not a bjd. @captain-exoplanet @cmccully could you confirm this? This has ~10 m/s implications for the BC correction.